src/main/io/gps.c

```diff
@@ -1337,60 +1340,105 @@ static void calculateNavInterval (void)
     gpsSol.navIntervalMs = constrain(navDeltaTimeMs, 50, 2500);
 }
 
+#if defined(USE_VIRTUAL_GPS)
+static void updateVirtualGPS(void)
+{
+    const uint32_t updateInterval = 100; // 100ms 10Hz update time interval
+    static uint32_t nextUpdateTime = 0;
+
+    if (cmp32(gpsData.now, nextUpdateTime) > 0) {
+        if (gpsData.state == GPS_STATE_INITIALIZED) {
+            gpsSetState(GPS_STATE_RECEIVING_DATA);
+        }
+
+        getVirtualGPS(&gpsSol);
+        gpsSol.time = gpsData.now;
+
+        gpsData.lastNavMessage = gpsData.now;
+        sensorsSet(SENSOR_GPS);
+
+        if (gpsSol.numSat > 3) {
+            gpsSetFixState(GPS_FIX);
+        } else {
+            gpsSetFixState(0);
+        }
+        GPS_update ^= GPS_DIRECT_TICK;
+
+        calculateNavInterval();
+        onGpsNewData();
+
+        nextUpdateTime = gpsData.now + updateInterval;
+    }
+}
+#endif
+
 void gpsUpdate(timeUs_t currentTimeUs)
 {
     static timeDelta_t gpsStateDurationFractionUs[GPS_STATE_COUNT];
     timeDelta_t executeTimeUs;
     gpsState_e gpsCurrentState = gpsData.state;
     gpsData.now = millis();
 
-    if (gpsPort) {
-        DEBUG_SET(DEBUG_GPS_CONNECTION, 7, serialRxBytesWaiting(gpsPort));
-        static uint8_t wait = 0;
-        static bool isFast = false;
-        while (serialRxBytesWaiting(gpsPort)) {
-            wait = 0;
-            if (!isFast) {
-                rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE_FAST));
-                isFast = true;
-            }
-            if (cmpTimeUs(micros(), currentTimeUs) > GPS_RECV_TIME_MAX) {
+    switch (gpsConfig()->provider) {
+        case GPS_UBLOX:
+        case GPS_NMEA:
```

Comment:
IIRC BF code recommends aligning `case` with `switch` (and I prefer that style in my code). gps.c uses both, but majority of `switch` statements indent `case`. So both variants are justified... 

```suggestion
    switch (gpsConfig()->provider) {
    case GPS_UBLOX:
    case GPS_NMEA:
```

Reply:
In this file  some switch have tab before case, some does not....

Reply:
Thank you! Will know about this rule.
But the next switch has tab before case too.
Seems, this file needs full switch code refactoring. 
Can i do this in this PR?

Reply:
Just adjust code that you are directly editing. We can fix all formatting in future, in single PR (and ideally enable lint checking at the some time).

Reply:
Ok